In [2]:
import pandas as pd

In [3]:
calendar = pd.read_csv("./data/calendar.txt")
calendar_dates = pd.read_csv("./data/calendar_dates.txt")

In [4]:
calendar.info()
calendar_dates.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44 entries, 0 to 43
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   service_id  44 non-null     int64
 1   monday      44 non-null     int64
 2   tuesday     44 non-null     int64
 3   wednesday   44 non-null     int64
 4   thursday    44 non-null     int64
 5   friday      44 non-null     int64
 6   saturday    44 non-null     int64
 7   sunday      44 non-null     int64
 8   start_date  44 non-null     int64
 9   end_date    44 non-null     int64
dtypes: int64(10)
memory usage: 3.6 KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 498 entries, 0 to 497
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   service_id      498 non-null    int64
 1   date            498 non-null    int64
 2   exception_type  498 non-null    int64
dtypes: int64(3)
memory usage: 11.8 KB


In [5]:
# 86832 -> tuesday , wednesday, thursday
# 57053 -> saturday
# 20769 -> sunday
service_ids = [86832,57053,20769]

In [6]:
trips = pd.read_csv("./data/trips.txt")

In [7]:
#  get trips for the service ids
reqTrips = trips[trips['service_id'].isin(service_ids)]

In [8]:
trips.shape
reqTrips.shape

(11536, 13)

In [9]:
# min and max trips
trip_counts = trips.groupby('service_id').agg(num_trips=('trip_id', 'count')).reset_index()


In [10]:
trip_counts.sort_values(by='num_trips', ascending=False, inplace=True)
print(trip_counts)

    service_id  num_trips
22       86832      11536
15       23519       8735
13       17361       7778
23       87621        305
1          789        229
2         1776        197
10       16572        143
14       22063        136
9        14796        112
20       64769        104
21       77519         98
7         9313         98
4         2750         67
11       17029         65
0          332         65
12       17209         61
8        11226         61
18       30003         11
3         2655          2
5         6925          2
6         7574          2
17       28104          2
19       37684          2
16       27761          1


In [11]:
trips[trips["service_id"] == 27761]

route_id  service_id    trip_id  \
23480    100254       27761  729861698   

                                    trip_headsign  \
23480  Shoreline South 148th St Station Lake City   

                                tts_trip_headsign  trip_short_name  \
23480  Shoreline South 148th St Station Lake City              NaN   

       direction_id  block_id  shape_id  peak_flag  fare_id  \
23480             0   7633920  11065007          0      101   

       wheelchair_accessible  bikes_allowed  
23480                      1              1

In [12]:
# get stops for the trips
stopsTime = pd.read_csv("./data/stop_times.txt")

In [13]:
stopsTime[stopsTime["trip_id"] == 729861698]

trip_id arrival_time departure_time  stop_id  stop_sequence  \
926780  729861698     15:46:00       15:46:00    26010              1   
926781  729861698     15:46:21       15:46:21    26020              5   
926782  729861698     15:46:42       15:46:42    26030              8   
926783  729861698     15:47:21       15:47:21    26050             13   
926784  729861698     15:48:00       15:57:00    26060             17   
926785  729861698     15:58:35       15:58:35    26080             22   
926786  729861698     16:00:14       16:00:14    26100             33   
926787  729861698     16:01:01       16:01:01    26110             35   
926788  729861698     16:02:00       16:02:00    10860             41   
926789  729861698     16:02:36       16:02:36    10870             44   
926790  729861698     16:03:27       16:03:27    35230             50   
926791  729861698     16:04:45       16:04:45    38670             53   
926792  729861698     16:05:31       16:05:31    38680             55   
926793  729861698     16:07:26       16:07:26    38700             58   
926794  729861698     16:09:06       16:09:06    38720             65   
926795  729861698     16:10:00       16:10:00    38730             68   
926796  729861698     16:11:01       16:11:01    82120             73   
926797  729861698     16:12:00       16:12:00    82130             76   
926798  729861698     16:13:03       16:13:03    82140             79   
926799  729861698     16:15:18       16:15:18    38810             87   
926800  729861698     16:16:56       16:16:56    82160             93   
926801  729861698     16:23:00       16:23:00    81301            123   

       stop_headsign  pickup_type  drop_off_type  shape_dist_traveled  \
926780           NaN            0              0                  0.0   
926781           NaN            0              0                693.5   
926782           NaN            0              0               1396.4   
926783           NaN            0              0               2688.1   
926784           NaN            0              0               3982.0   
926785           NaN            0              0               5271.8   
926786           NaN            0              0               6623.3   
926787           NaN            0              0               7272.6   
926788           NaN            0              0               8076.3   
926789           NaN            0              0               8515.7   
926790           NaN            0              0               9149.6   
926791           NaN            0              0              10104.8   
926792           NaN            0              0              10668.1   
926793           NaN            0              0              12090.1   
926794           NaN            0              0              13321.1   
926795           NaN            0              0              13982.6   
926796           NaN            0              0              14673.5   
926797           NaN            0              0              15331.1   
926798           NaN            0              0              16044.7   
926799           NaN            0              0              17568.2   
926800           NaN            0              0              18667.5   
926801           NaN            0              0              22779.0   

        timepoint  
926780          1  
926781          0  
926782          0  
926783          0  
926784          1  
926785          0  
926786          0  
926787          0  
926788          1  
926789          0  
926790          0  
926791          0  
926792          0  
926793          0  
926794          0  
926795          1  
926796          0  
926797          0  
926798          0  
926799          0  
926800          0  
926801          1

In [14]:
calendar[calendar['service_id'] == 27761]

service_id  monday  tuesday  wednesday  thursday  friday  saturday  \
14       27761       0        1          0         1       1         0   

    sunday  start_date  end_date  
14       0    20250501  20250617

In [15]:
calendar_dates[calendar_dates['service_id'] == 27761]

service_id      date  exception_type
319       27761  20250526               2
320       27761  20250609               1
321       27761  20250602               1
322       27761  20250519               1
323       27761  20250512               1
324       27761  20250616               1
325       27761  20250505               1

<!-- generated test case -->
<!-- single trip id -->

{
    title: "valid trip day (19 may 25)"
    service id : 27761
    days : {tuesday , thursday , friday} 
    <!-- see exceptions -->
    trip_id : 729861698
    stop_id : 26080
    system_time = 1747695600000

    result : {
        trips : {
            1_729861698,
            1_724948468,
            1_724948328
        }
    }
}

{
    title: "non service trip day ()"
    service id : 27761
    days : {tuesday , thursday , friday} 
    <!-- see exceptions -->
    trip_id : 729861698
    stop_id : 26080
    system_time = 1747868400000

    result : {
        trips : {
            1_724948468,
            1_724948328
        }
    }
}

In [16]:
gs = stopsTime.groupby('stop_id').agg(num_of_trips=('trip_id', 'count')).reset_index().sort_values(by='num_of_trips', ascending=False)

In [17]:
gs

stop_id  num_of_trips
52        600          2787
32        420          2685
50        575          2544
48        558          2368
45        538          2001
...       ...           ...
4921    70588             1
4906    70528             1
4909    70533             1
6112    82718             1
1834    31135             1

[6422 rows x 2 columns]

In [18]:
gs[gs["stop_id"] == 26080]

stop_id  num_of_trips
1554    26080           211

In [19]:
stopsTime.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1037691 entries, 0 to 1037690
Data columns (total 10 columns):
 #   Column               Non-Null Count    Dtype  
---  ------               --------------    -----  
 0   trip_id              1037691 non-null  int64  
 1   arrival_time         1037691 non-null  object 
 2   departure_time       1037691 non-null  object 
 3   stop_id              1037691 non-null  int64  
 4   stop_sequence        1037691 non-null  int64  
 5   stop_headsign        213014 non-null   object 
 6   pickup_type          1037691 non-null  int64  
 7   drop_off_type        1037691 non-null  int64  
 8   shape_dist_traveled  1037691 non-null  float64
 9   timepoint            1037691 non-null  int64  
dtypes: float64(1), int64(6), object(3)
memory usage: 79.2+ MB


In [20]:
stop26080Trips =  stopsTime[stopsTime["stop_id"] == 26080].sort_values(by="arrival_time", inplace=False)

In [21]:
# Step 1: Filter stop_times for stop_id 26080
stop26080Trips = stopsTime[stopsTime["stop_id"] == 26080].sort_values(by="arrival_time", inplace=False)

# Step 2: Merge with trips to get service_id
stop26080Trips = stop26080Trips.merge(trips[['trip_id', 'service_id']], on='trip_id', how='left')

# Step 3: Merge with calendar to get day-of-week activity
stop26080Trips = stop26080Trips.merge(calendar, on='service_id', how='left')

# Step 4 (Optional): Select and reorder relevant columns
days_columns = ['monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday']
cols_to_show = ['trip_id', 'arrival_time', 'departure_time', 'stop_id'] + days_columns

stop26080Trips = stop26080Trips[cols_to_show]

stop26080Trips.to_csv("./data/arrival_departures_for_stop26080Trips.csv", index=False)


In [22]:
trips[trips["trip_id"] == 571921968]

route_id  service_id    trip_id        trip_headsign  \
4100    100345       30003  571921968  Haller Lake Special   

        tts_trip_headsign  trip_short_name  direction_id  block_id  shape_id  \
4100  haller lake special              NaN             1   7633493  10987002   

      peak_flag  fare_id  wheelchair_accessible  bikes_allowed  
4100          0      101                      1              1

In [25]:
routes = pd.read_csv("./data/routes.txt")
routes[routes["route_id"] == 100345]

route_id  agency_id route_short_name  route_long_name  \
90    100345          1              987              NaN   

                         route_desc  route_type  \
90  Rainier Beach - Lakeside School           3   

                                            route_url route_color  \
90  https://kingcounty.gov/en/dept/metro/routes-an...      FDB71A   

   route_text_color  
90           000000

In [23]:
calendar_dates[calendar_dates['service_id'] == 30003]

service_id      date  exception_type
270       30003  20250526               2
271       30003  20250602               1
272       30003  20250519               1
273       30003  20250512               1
274       30003  20250505               1

In [24]:
calendar[calendar['service_id'] == 30003]

service_id  monday  tuesday  wednesday  thursday  friday  saturday  \
11       30003       0        1          1         1       1         0   

    sunday  start_date  end_date  
11       0    20250501  20250606

trip 571921968 is serviced on weekdays except mon from 1 may to 6 june and few exceptions 
service_id	date	exception_type
270	30003	20250526	2 //"mon off" no need already off?
271	30003	20250602	1
272	30003	20250519	1
273	30003	20250512	1
274	30003	20250505	1
